### ISSUE TITLE
Double quotes in filter - no result

### ISSUE THREAD

**Issue by mrmaxonline** [meilisearch/meilisearch#1653]
Value for filter with double quotes return no result.

Filter sample: d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = "ООО \\"ХАРДЭНЕРДЖИ\\""
Q is empty = no result.

In web simplte search result present:

![image](https://user-images.githubusercontent.com/12570364/131450757-9bde0084-fe0d-440e-9940-147d2f0364fd.png)

![image](https://user-images.githubusercontent.com/12570364/131450849-75ce3714-cf49-45c2-adf7-d56540ed89ca.png)

MeiliSearch version: v0.21.0

**Comment by curquiza** [comment]
Hello @mrmaxonline!
What is the exact request, including the `filter` parameter you send to MeiliSearch?
You need to configure and use `filter` in a specific way when filtering, see [our dedicated page on the docs](https://docs.meilisearch.com/reference/features/filtering_and_faceted_search.html#configuring-filters)

**Comment by mrmaxonline** [comment]
Hi @curquiza !
Filter is configured and work fine wihout double quotes inside the string (value of filter).

SearchQuery: {"AttributesToCrop":null,"AttributesToHighlight":["*"],"AttributesToRetrieve":null,"CropLength":null,"FacetFilters":null,"FacetsDistribution":null,"Filter":"d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = \"ООО \\\"ХАРДЭНЕРДЖИ\\\"\"","Limit":100,"Matches":null,"Offset":0,"Q":null}
Filters: d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = "ООО \"ХАРДЭНЕРДЖИ\""
Page Index: 0, Page Size: 100
Result count: 0

And good result without double quotes in filter is:
SearchQuery: {"AttributesToCrop":null,"AttributesToHighlight":["*"],"AttributesToRetrieve":null,"CropLength":null,"FacetFilters":null,"FacetsDistribution":null,"Filter":"d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = \"ТЕХСУ ООО\"","Limit":100,"Matches":null,"Offset":0,"Q":null}
Filters: d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = "ТЕХСУ ООО"
Page Index: 0, Page Size: 100
Result count: 28

**Comment by curquiza** [comment]
Can you give us the document please, with the whole attributes, so that I can copy/paste and try to reproduce please? :)
Plus, which SDK are you using? I'm curious since there is still the `FacetFilters` that should not be present anymore

**Comment by mrmaxonline** [comment]
`{"documentUid":"c32fe55f-cd1e-43e6-90f9-e8c956fa4118","changeTimestamp":1630396821,"typeUid":"1a8745a8-2b6e-4f4c-a7d2-ee5f175a7e8b","source":1,"direction":1,"creationDate":"2021-08-30T19:51:43+03:00","changeDate":"2021-08-31T08:00:21.7435326+03:00","status":"Документооборот завершен","author":null,"name":"Счет №1264 от 30.08.21","docType":2,"infraObjects":null,"fieldUids":["d8265ba0-8628-4d61-bb3b-3d6c770027af","c6b56f17-1387-40f8-93be-b4e0f5e02ec6","3463cc4d-7410-4037-ac5f-62c15f3b8f55"],"fieldValues":["ООО \"Русские Башни\"","ООО \"ХАРДЭНЕРДЖИ\"","Счет на оплату"],"hasMapping":true,"warningMessage":null,"textBody":null,"haveFile":true,"d_9250ae64-6b04-4c94-a647-222203f3d267":null,"d_b5d72d9e-8ba5-485c-bd28-958d7164db75":"31.08.2021","d_a4adc938-6c4a-4952-8728-faf37e43adee":"30.08.2021","d_9f4e546a-f953-4d8e-b908-cec80decb69e":null,"d_3206381b-20b3-4101-9e66-ae7ecc78b54a":null,"d_ee7c8fe9-64b2-4fdf-9658-50c0e8d8fd2d":null,"d_6ca7820b-687e-4d45-b894-6f0140f05323":null,"d_33f27361-1f16-47a1-9082-3433be2723e5":null,"d_4fb5a080-70ff-4442-8e51-e5da3b62d73c":null,"d_40eb24f9-c36e-4ca9-bb38-089d1e9eb74e":null,"d_3463cc4d-7410-4037-ac5f-62c15f3b8f55":"Счет на оплату","d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6":"ООО \"ХАРДЭНЕРДЖИ\"","d_d8265ba0-8628-4d61-bb3b-3d6c770027af":"ООО \"Русские Башни\"","d_9baf57ce-3a93-4191-8eb4-d330f7a696b4":"Документооборот завершен"}`

**Comment by mrmaxonline** [comment]
> * Plus, which SDK are you using? I'm curious since there is still the `FacetFilters` that should not be present anymore

Fixed in code.

**Comment by curquiza** [comment]
Thanks a lot for sharing this!

We think there is indeed a bug when using `"` if the document already contains `"`, despite the escaping.

### Workaround

Use `'` for literal strings instead of `"`.
In your case
```json
{
  "q": "",
  "filter": "d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'"
}
```

### Bug

To help the @meilisearch/core-team to reproduce

- Add this document:
```json
[
  { "id": 2, "title": "Pride and Prejudice", "author": "Jane \"Austin\"" }
]
```
- Set `author` as `filterableAttributes`
- Search for:
```json
{
    "filter": "author = 'Jane \"Austin\"'"
}
```
-> Ok, I get a result.
- Search for:
```json
{
    "filter": "author = \"Jane \\\"Austin\\\"\""
}
```
-> I don't get any result despite the escape of characters

**Comment by mrmaxonline** [comment]
> ### Workaround
> Use `'` for literal strings instead of `"`.
> In your case
> 
> ```json
> {
>   "q": "",
>   "filter": "d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'"
> }
> ```

Hm... error:

`MeilisearchApiError, Message:  --> 1:42
  |
1 | d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'
  |                                          ^---
  |
  = expected word, ErrorCode: invalid_filter, ErrorType: invalid_request_error, ErrorLink: https://docs.meilisearch.com/errors#invalid_filter
`

**Comment by curquiza** [comment]
Hum, weird. With 
```
{
  "q": "",
  "filter": "d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'"
}
```

I get:
```
{
    "hits": [
        {
            "documentUid": "c32fe55f-cd1e-43e6-90f9-e8c956fa4118",
            "changeTimestamp": 1630396821,
            "typeUid": "1a8745a8-2b6e-4f4c-a7d2-ee5f175a7e8b",
            "source": 1,
            "direction": 1,
            "creationDate": "2021-08-30T19:51:43+03:00",
            "changeDate": "2021-08-31T08:00:21.7435326+03:00",
            "status": "Документооборот завершен",
            "author": null,
            "name": "Счет №1264 от 30.08.21",
            "docType": 2,
            "infraObjects": null,
            "fieldUids": [
                "d8265ba0-8628-4d61-bb3b-3d6c770027af",
                "c6b56f17-1387-40f8-93be-b4e0f5e02ec6",
                "3463cc4d-7410-4037-ac5f-62c15f3b8f55"
            ],
            "fieldValues": [
                "ООО \"Русские Башни\"",
                "ООО \"ХАРДЭНЕРДЖИ\"",
                "Счет на оплату"
            ],
            "hasMapping": true,
            "warningMessage": null,
            "textBody": null,
            "haveFile": true,
            "d_9250ae64-6b04-4c94-a647-222203f3d267": null,
            "d_b5d72d9e-8ba5-485c-bd28-958d7164db75": "31.08.2021",
            "d_a4adc938-6c4a-4952-8728-faf37e43adee": "30.08.2021",
            "d_9f4e546a-f953-4d8e-b908-cec80decb69e": null,
            "d_3206381b-20b3-4101-9e66-ae7ecc78b54a": null,
            "d_ee7c8fe9-64b2-4fdf-9658-50c0e8d8fd2d": null,
            "d_6ca7820b-687e-4d45-b894-6f0140f05323": null,
            "d_33f27361-1f16-47a1-9082-3433be2723e5": null,
            "d_4fb5a080-70ff-4442-8e51-e5da3b62d73c": null,
            "d_40eb24f9-c36e-4ca9-bb38-089d1e9eb74e": null,
            "d_3463cc4d-7410-4037-ac5f-62c15f3b8f55": "Счет на оплату",
            "d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6": "ООО \"ХАРДЭНЕРДЖИ\"",
            "d_d8265ba0-8628-4d61-bb3b-3d6c770027af": "ООО \"Русские Башни\"",
            "d_9baf57ce-3a93-4191-8eb4-d330f7a696b4": "Документооборот завершен"
        }
    ],
    "nbHits": 1,
    "exhaustiveNbHits": false,
    "query": "",
    "limit": 20,
    "offset": 0,
    "processingTimeMs": 1
}
```

How do you send you filter?

**Comment by mrmaxonline** [comment]
````
SearchQuery: {"AttributesToCrop":null,"AttributesToHighlight":["*"],"AttributesToRetrieve":null,"CropLength":null,"FacetsDistribution":null,"Filter":"d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \\\"ХАРДЭНЕРДЖИ\\\"'","Limit":100,"Matches":null,"Offset":0,"Q":null}
````

**Comment by mrmaxonline** [comment]
Fixed!

````
SearchQuery: {"AttributesToCrop":null,"AttributesToHighlight":["*"],"AttributesToRetrieve":null,"CropLength":null,"FacetsDistribution":null,"Filter":"d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'","Limit":100,"Matches":null,"Offset":0,"Q":null}
````
Result count: 100

**Comment by curquiza** [comment]
> Thanks a lot for sharing this!
> 
> We think there is indeed a bug when using `"` if the document already contains `"`, despite the escaping.
> ### Workaround
> 
> Use `'` for literal strings instead of `"`. In your case
> 
> ```json
> {
>   "q": "",
>   "filter": "d_c6b56f17-1387-40f8-93be-b4e0f5e02ec6 = 'ООО \"ХАРДЭНЕРДЖИ\"'"
> }
> ```
> 
> ### Bug
> 
> To help the @meilisearch/core-team to reproduce
> 
>     * Add this document:
> 
> 
> ```json
> [
>   { "id": 2, "title": "Pride and Prejudice", "author": "Jane \"Austin\"" }
> ]
> ```
> 
>     * Set `author` as `filterableAttributes`
> 
>     * Search for:
> 
> 
> ```json
> {
>     "filter": "author = 'Jane \"Austin\"'"
> }
> ```
> 
> -> Ok, I get a result.
> 
>     * Search for:
> 
> 
> ```json
> {
>     "filter": "author = \"Jane \\\"Austin\\\"\""
> }
> ```
> 
> -> I don't get any result despite the escape of characters

Current state in v0.25.0 (including the new parsing with nom instead of pest). We get an error message and not an empty result.

```json
{
    "message": "Found unexpected characters at the end of the filter: `Austin\\\\\\\"\\\"`. You probably forgot an `OR` or an `AND` rule.\n18:27 author = \"Jane \\\"Austin\\\"\"",
    "code": "invalid_filter",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#invalid_filter"
}
```

Maybe do I do something wrong. Do you have an idea @irevoire?

**Comment by curquiza** [comment]
Fixed by https://github.com/meilisearch/meilisearch/pull/2005 that bump the milli dependency to [milli v0.22.0](https://github.com/meilisearch/milli/releases/tag/v0.22.0) containing the fix of this issue 🙂

The bug fix will be released in Meilisearch v0.26.0.
